### ISSUE TITLE
SAW crashes with error "saveLocals: ..."

### ISSUE THREAD

**Issue by nano-o** [GaloisInc/saw-script#433]
The following script produces:
```
saveLocals:
        [0,2,3,4,6,8,10,12]
        [0,2,3]
CallStack (from HasCallStack):
  error, called at src/Lang/Crucible/JVM/Generator.hs:197:15 in crucible-jvm-0.2-F1V0Hxopa9dJFsMKd3xh5f:Lang.Crucible.JVM.Generator
  jvmFail, called at src/Lang/Crucible/JVM/Translation.hs:406:5 in crucible-jvm-0.2-F1V0Hxopa9dJFsMKd3xh5f:Lang.Crucible.JVM.Translation                                                                                                                                       
  saveLocals, called at src/Lang/Crucible/JVM/Translation.hs:417:6 in crucible-jvm-0.2-F1V0Hxopa9dJFsMKd3xh5f:Lang.Crucible.JVM.Translation                                                                                                                                    
  writeRegisters, called at src/Lang/Crucible/JVM/Translation.hs:469:13 in crucible-jvm-0.2-F1V0Hxopa9dJFsMKd3xh5f:Lang.Crucible.JVM.Translation                                                                                                                               
  processBlockAtPC, called at src/Lang/Crucible/JVM/Translation.hs:949:24 in crucible-jvm-0.2-F1V0Hxopa9dJFsMKd3xh5f:Lang.Crucible.JVM.Translation                                                                                                                             
  generateInstruction, called at src/Lang/Crucible/JVM/Translation.hs:442:38 in crucible-jvm-0.2-F1V0Hxopa9dJFsMKd3xh5f:Lang.Crucible.JVM.Translation                                                                                                                          
```

```
enable_experimental;
import "../cryptol/Encryption/Arith.cry" as arith;

myclass <- java_load_class "com.galois.crypto.Arith";

let jvm_array_is ref x =
  do {
    let xs = eval_list x;
    let l = length xs;
    let is = eval_list {{ [0..(l-1)]:[_][width (l-1)] }};
    for is (\i -> jvm_elem_is ref (eval_int i) (jvm_term (nth xs (eval_int i))));
    return ();
  };

let method_prove name specs setup tac =
  time do {
    print (str_concat "Verifying " name);
    crucible_jvm_verify myclass name specs false setup tac;
  };


let divide_java_setup = 
  do {
    u <- (jvm_fresh_var "u" java_long);
    v <- (jvm_fresh_var "v" java_int);
    jvm_precond {{ ~(v@0) && (v != 0) }};

    jvm_execute_func [jvm_term u, jvm_term v];

    ret <- jvm_alloc_array 2 java_long;
    jvm_array_is ret {{ arith::divide_java u v }};

    jvm_return ret;
  };

divide_java_ov <- method_prove "divideUnsigned" [] divide_java_setup z3;
```
```
divide_java : {bw} (fin bw, bw >= 1) => [2*bw] -> [bw] -> [2][2*bw]
divide_java u v = if (v == 1) then [u,0] else res
  where 
    vd = (zero#v):[2*bw]
    u' = u >> 1 // eliminates sign bit
    v' = vd >> 1
    qh = u'/v' // estimate quotient
    rh2 = (u'%v')<<1 // estimate remainder
    res = // now correct the estimates
        if ((u!0) && ~(v!0))
        then [qh, rh2+1]
        else
          if (~(u!0) && (v!0))
          then
            if (rh2 < qh)
            then [qh-c, rh2-qh+(c*vd)]
            else [qh, rh2-qh]
              where c = (ceil_div (qh-rh2) vd)
          else
            if ((u!0) && (v!0))
            then 
              if (rh2+1 < qh)
              then [qh-c, rh2-qh+1+(c*vd)]
              else [qh, rh2-qh+1]
                where c = (ceil_div (qh-rh2-1) vd)
            else [qh,rh2]

ceil_div : {bw} (fin bw, bw >= 1) => [bw] -> [bw] -> [bw]
ceil_div a b = if (a%b == 0) then a/b else (a/b)+1
```
```
package com.galois.crypto;

public class Arith {

  public static long ceilDiv(long a, long b) {
	  if (a%b == 0)
		  return a/b;
	  else 
		  return a/b+1;
  }
  
  public static long[] divideUnsigned(long u, int v) {
	  long[] res = new long[2];
	  if (v == 1) {
		  res[0] = u;
		  res[1] = 0L;
	  }
	  else {
		  long vd = (long)v;
		  long u1 = u >>> 1;
		  long v1 = vd >>> 1;
		  long qh = u1/v1; // both u1 and v1 have their sign bit off so signed division and unsigned division are the same. Also note that qh has a 0 msb here.
		  long rh2 = (u1%v1)<<1;
		  if ((u & 1L) > 0 && (v & 1) == 0) {
			  res[0] = qh;
			  res[1] = rh2+1;
		  }
		  else if ((u & 1L) == 0 && (v & 1) > 0) {
			  long rh = rh2 - qh;
			  long c = ceilDiv(-rh, vd);
			  if (rh2 < qh) {
				  res[0] = qh - c;
				  res[1] = rh + c*vd;
			  }
			  else {
				  res[0] = qh;
				  res[1] = rh;
			  }
		  }
		  else if ((u & 1L) > 0 && (v & 1) > 0) {
			  long rh = rh2-qh+1;
			  long c = ceilDiv(-rh, vd);
			  if (rh2+1 < qh) {
				  res[0] = qh - c;
				  res[1] = rh + c*vd;
			  }
			  else {
				  res[0] = qh ;
				  res[1] = rh;
			  }
		  }
		  else {
			  res[0] = qh;
			  res[1] = rh2;
		  }
	  }
	  return res;
  }
}
```

**Comment by brianhuffman** [comment]
I see what is going on here. The problem has to do with the way that `crucible-jvm` translates JVM code into crucible; specifically with how it determines the set of inputs for each basic block. Translation starts at the method entry point, and then whenever we have a jump to a new basic block for the first time, we decide the inputs for that block based on what local values we have available to pass in. When we see a branch to an already-seen basic block, we look in a table to learn what inputs it expects.

To store local values, the JVM uses both a stack and a table of local variables. It seems that when there is more than one code path leading to the same basic block, the Java compiler always ensures that the stack shape will always be the same at that point. So each basic block can have an input for each element on the stack, and no matter where we come from, we will always have values available for all of those input slots.

However, it seems that the same is not true for the local variable table. In the example in the OP, one code path arrives at a particular basic block with entries [0,2,3,4,6,8,10,12] in the local variable table filled, so we define that block to have eight corresponding inputs. Later, another code path jumps to the same basic block when the local variable table only has entries [0,2,3]. Now we can't provide all the inputs that we defined the block to have.

It turns out that it was a mistake to define that block as having eight inputs for local variables; in fact the block must have actually ignored all the local variables besides [0,2,3], so we should have only defined three inputs. Instead of making an input for every local variable in the table, we should only make an input for each local variable that is actually used in that block.

**Comment by brianhuffman** [comment]
Fixed by GaloisInc/crucible#231.
